Name: <redacted> <redacted>

GitHub Username: <redacted>

GitHub Repo: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

Commit ID: 1361796a5e3b8b885a46cc00209149dc32ff31dc

In [25]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import *
from sklearn.linear_model import *
from sklearn.preprocessing import *
from sklearn.svm import *
from sklearn.neural_network import *
from sklearn.neighbors import *

# Part 1: Load the dataset

In [26]:
# Load the dataset (load remotely or locally)
iris_dataset = pd.read_csv("iris.csv")
# Output the first 15 rows of the data
print("Iris Dataset:")
print(iris_dataset.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print("\nSummary:")
print(iris_dataset.info())

unique_species = iris_dataset['species'].unique()
print("\nUnique Species:")
print(unique_species)


Iris Dataset:
    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data of 150 entries gives the lengths and widths of both the sepal and petals of 3 different kinds of similarly looking flowers. There are 4 features: sepal_length, sepal_width, petal_length, and petal_width. The label is what kind of species of flower each row of data corresponds to and is found in the "species" column. There are three different species which are setosa, versicolor, and virginica. 

# Part 2: Split the dataset into train and test

In [27]:
# Take the dataset and split it into our features (X) and label (y)
features = iris_dataset[["sepal_length", "sepal_width", "petal_length", "petal_width"]]
label = iris_dataset["species"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.3, train_size=0.7, random_state=13)

# Part 3: Logistic Regression

In [28]:
# i. Use sklearn to train a LogisticRegression model on the training set
logistic_model = LogisticRegression(max_iter=200)
logistic_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data_point = pd.DataFrame(X_test.iloc[0].values.reshape(1, -1), columns=X_train.columns)
logistic_probabilities = logistic_model.predict_proba(sample_data_point)
class_labels = logistic_model.classes_

print("Predicted probabilities for each class:")
for class_label, probability in zip(class_labels, logistic_probabilities[0]):
    print(f"{class_label}: {probability:.5f}")

# iii. Report on the score for Logistic regression model, what does the score measure?
accuracy_score = logistic_model.score(X_test, y_test)
print(f"\nAccuracy score of the logistic regression model: {accuracy_score}")

# iv. Extract the coefficents and intercepts for the boundary line(s)
feature_names = features.columns
coefficients_df = pd.DataFrame(logistic_model.coef_, columns=feature_names, index=class_labels)
print("\nCoefficients for each feature:")
print(coefficients_df)

intercept_df = pd.DataFrame(logistic_model.intercept_, index=class_labels, columns=['Intercept'])
print("\nIntercepts for each class:")
print(intercept_df)

# Use the cross_val_score function to repeat your experiment across many shuffles of the data
reg_cv_score = cross_val_score(logistic_model, features, label, cv=10)
print("\nCross-validation scores:")
print(reg_cv_score)

# Report on their finding and their significance
print("\nMean:")
print(reg_cv_score.mean())
print("\nStandard deviation:")
print(reg_cv_score.std())

Predicted probabilities for each class:
setosa: 0.00807
versicolor: 0.83578
virginica: 0.15615

Accuracy score of the logistic regression model: 0.9777777777777777

Coefficients for each feature:
            sepal_length  sepal_width  petal_length  petal_width
setosa         -0.379876     0.881450     -2.293558    -0.988085
versicolor      0.589672    -0.317787     -0.154278    -0.715928
virginica      -0.209796    -0.563663      2.447835     1.704013

Intercepts for each class:
            Intercept
setosa       8.768576
versicolor   1.259487
virginica  -10.028064

Cross-validation scores:
[1.         0.93333333 1.         1.         0.93333333 0.93333333
 0.93333333 1.         1.         1.        ]

Mean:
0.9733333333333334

Standard deviation:
0.03265986323710904


The score function for the logistic regression model returns the mean accuracy on the given test data and labels. In multi-label classification, this is the subset accuracy which is a harsh metric since you require for each sample that each label set be correctly predicted. The score of 0.97778 for the regression model means that the model is performing very well and is correctly classifying about 97% of the data points.

# Part 4: Support Vector Machine

In [29]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = SVC(probability=True)
svm_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
svm_probabilities = svm_model.predict_proba(sample_data_point)

print("Predicted probabilities for each class:")
for class_label, probability in zip(class_labels, svm_probabilities[0]):
    print(f"{class_label}: {probability:.5f}")

# iii. Report on the score for the SVM, what does the score measure?
svm_score = svm_model.score(X_test, y_test)
print(f"\nAccuracy score of the Support Vector Classifier: {svm_score}")

# Use the cross_val_score function to repeat your experiment across many shuffles of the data
svm_cv_score = cross_val_score(svm_model, features, label, cv=10)
print("\nCross-validation scores:")
print(svm_cv_score)

# Report on their finding and their significance
print("\nMean:")
print(svm_cv_score.mean())
print("\nStandard deviation:")
print(svm_cv_score.std())

Predicted probabilities for each class:
setosa: 0.00804
versicolor: 0.92641
virginica: 0.06554

Accuracy score of the Support Vector Classifier: 0.9555555555555556

Cross-validation scores:
[1.         0.93333333 1.         1.         1.         0.93333333
 0.93333333 0.93333333 1.         1.        ]

Mean:
0.9733333333333334

Standard deviation:
0.03265986323710904


The accuracy score measures the proportion of correctly predicted instances to the total instances in the test set. Since the SVM model got a score of 0.95556, it means the model is performing pretty well and is correctly predicting the class of about 96% of the data points in the test set.

# Part 5: Neural Network

In [32]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(max_iter=1000, random_state=42)
mlp_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
mlp_probabilities = mlp_model.predict_proba(sample_data_point)

print("Predicted probabilities for each class:")
for class_label, probability in zip(class_labels, mlp_probabilities[0]):
    print(f"{class_label}: {probability:.5f}")

# iii. Report on the score for the Neural Network, what does the score measure?
mlp_score = mlp_model.score(X_test, y_test)
print(f"\nAccuracy score of the Neural Network: {mlp_score}\n")

# Use the cross_val_score function to repeat your experiment across many shuffles of the data
mlp_cv_score = cross_val_score(mlp_model, features, label, cv=10)
print("\nCross-validation scores:")
print(mlp_cv_score)

# Report on their finding and their significance
print("\nMean:")
print(mlp_cv_score.mean())
print("\nStandard deviation:")
print(mlp_cv_score.std())
print()

# iv: Experiment with different options for the neural network, report on your best configuration
best_accuracy = 0
best_config = None
configurations = [
    {"hidden_layer_sizes": (10,), "activation": "relu"},
    {"hidden_layer_sizes": (20,), "activation": "relu"},
    {"hidden_layer_sizes": (10, 10), "activation": "relu"},
    {"hidden_layer_sizes": (20, 20), "activation": "relu"},
    {"hidden_layer_sizes": (10,), "activation": "logistic"},
    {"hidden_layer_sizes": (10, 10), "activation": "logistic"},
    {"hidden_layer_sizes": (30, 30, 30), "activation": "tanh"}
]

for config in configurations:
    mlp_model = MLPClassifier(hidden_layer_sizes=config["hidden_layer_sizes"],
                               activation=config["activation"],
                               max_iter=10000,
                               random_state=42)
    mlp_model.fit(X_train, y_train)
    score = mlp_model.score(X_test, y_test)
    print(f"Configuration: {config}, Score: {score:.5f}")
    
    if score > best_accuracy:
        best_accuracy = score
        best_config = config

print(f"\nBest Configuration: {best_config} with Accuracy Score: {best_accuracy:.5f}")

Predicted probabilities for each class:
setosa: 0.00072
versicolor: 0.94623
virginica: 0.05305

Accuracy score of the Neural Network: 0.9777777777777777


Cross-validation scores:
[1.         1.         1.         1.         0.86666667 1.
 0.93333333 1.         1.         1.        ]

Mean:
0.9800000000000001

Standard deviation:
0.04268749491621898

Configuration: {'hidden_layer_sizes': (10,), 'activation': 'relu'}, Score: 0.97778
Configuration: {'hidden_layer_sizes': (20,), 'activation': 'relu'}, Score: 0.97778
Configuration: {'hidden_layer_sizes': (10, 10), 'activation': 'relu'}, Score: 0.97778
Configuration: {'hidden_layer_sizes': (20, 20), 'activation': 'relu'}, Score: 0.97778
Configuration: {'hidden_layer_sizes': (10,), 'activation': 'logistic'}, Score: 1.00000
Configuration: {'hidden_layer_sizes': (10, 10), 'activation': 'logistic'}, Score: 0.97778
Configuration: {'hidden_layer_sizes': (30, 30, 30), 'activation': 'tanh'}, Score: 0.95556

Best Configuration: {'hidden_layer_sizes'

Similarly to the above two classifying models, the neaural network also got a score of 0.97778 which means that it is correctly classifying almost every data point in the test set. This means the model is well trained for this data.

With the testing of a few more neural networks, the best performing setup seemed to be when the hidden layer size was 10 and the activation was set to logistic. This model performed at the highest level and got a score of 1.0 which means that it was able to classify 100% of the data correctly which is very impressive.

# Part 6: K-Nearest Neighbors

In [31]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
k = 3
knn_model = KNeighborsClassifier(n_neighbors=k)
knn_model.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
knn_probabilities = knn_model.predict_proba(sample_data_point)
for class_label, probability in zip(class_labels, knn_probabilities[0]):
    print(f"{class_label}: {probability:.5f}")

# iii. Report on the score for kNN, what does the score measure?
knn_score = knn_model.score(X_test, y_test)
print(f"\nAccuracy score of the k-Neighbors Classifier: {knn_score}")

# Use the cross_val_score function to repeat your experiment across many shuffles of the data
knn_cv_score = cross_val_score(knn_model, features, label, cv=10)
print("\nCross-validation scores:")
print(knn_cv_score)

# Report on their finding and their significance
print("\nMean:")
print(knn_cv_score.mean())
print("\nStandard deviation:")
print(knn_cv_score.std())

setosa: 0.00000
versicolor: 1.00000
virginica: 0.00000

Accuracy score of the k-Neighbors Classifier: 0.9555555555555556

Cross-validation scores:
[1.         0.93333333 1.         0.93333333 0.86666667 1.
 0.93333333 1.         1.         1.        ]

Mean:
0.9666666666666666

Standard deviation:
0.04472135954999579


The KNN model has the same lowest score as the SVM. The score is 0.95556 and while it is correctly classifying about 96% of the data, it is still performing at a lower rate than logistic regression and the neural network.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

It took a lot of manipulation of the dataset to finally see some differences in the scores of the models. When I changed the random state of the train-test split, all of the scores would change in the same way. At first I used a random state of 42 and a train-test split of 90% and 10%, respectively, however, that made all of the models have a score of 1 which meant they were all performing perfectly by assigning the classes of the test data correctly every time.

I also attempted to look at cross-validation scores as well, however, similarly to the accuracy scores, there just wasn't enough of a difference between the mean scores of the cross-validation sets. However, using cross validation, some of the models did recieve a significantly lower individual score within the set of many 1.0s which I will talk about later in this conclusion.

In order to get a variance in the scores, I decided to first change the random state of the split. However, like I said above, this would change each score very similarly. Therefore, I decided to also change the percentage of data that was allocated to the training set and testing set to 80% and 20%, respectively. I wasn't seeing as much variety in scores as I hoped from this and so I changed the percentages again to 70% and 30%. Here is where I saw more easily interpretable scores.

The logistic regression model saw a score of 0.97778. The cross validation verified that the score was relatively accurate as the mean of the cross validation was 0.97334 with one of the smallest standard deviations of 0.03266 meaning the scores were closely clustered around the mean and that score is a good indicator of the true score. Over 10 iterations, the cross validation scored the model at 1.0 six times and at a minimum of 0.93333 four times.

The support vector machine saw a score of 0.95556. The cross validation actually came out with the same exact numbers as the logistic regression model above. This means the cross validation's mean gave this model a higher score at 0.97334. 

The neural network reported a score of 0.97778, same as the logistic regression model. However, the mean of the cross validation set was 0.98 which makes this the highest performing model for the 70/30 train-test split and the random seed of 13. While it did recieve a score of 1.0 for 8 out of the 10 trials, it also saw the lowest score out of all of the cross validation sets so far; that being 0.86667 for one of the trials.

The K-Nearest Neighbors model got a score of 0.95556, exactly the same as the SVM score. This model received the lowest cross validation mean at 0.96667 even though it is higher than the reported score of the model. It also got the same minimum score as the neaural network in its cross validation set of 0.86667. However, what brough the score down even more was the fact that it also saw a score of 0.93333 three times.

From the observations of the scores and cross validation sets, we can see a few interesting things. While the neural network did recieve one of the lowest cross validation scores, it also had the highest mean and the most amount of 1.0 scores and seemed to have performed the best on this train-test split. The logistic regression model appeared to have performed the second best as it had the same accuracy score as the neural model and a cross validation mean to support that score. You could make the case that the KNN model performed the worst from the fact that it had not only one of the worst scores, but also the worst cross validation mean. 

As for the reason why the neural network performed the best, it is most likely due to the fact that it can essentially model any continuous function which means that out of all of the different models tested here, it can most easily classify the data. As to why the KNN model might be worst performing, it is really no more than a way to guess where to classify the data rather than using an actual algorithm. I also found that the more neighbors you decide to look at, the worse the model performed in the case of this dataset. Therefore, it is not entirely suprising that the best performing would be the neural network and the worst performing is the K-nearest neighbors model.

I found it interesting that such small changes to the neural network configurations such as different size hidden layers or different amounts of hidden layers as well as the type of activation would change how well the model performed. It was cool to see the one and only perfectly scored model for this homework was a neural network with one size 10 hidden layer and a logistic activation.